doc/tutorial.rst

```diff
@@ -0,0 +1,317 @@
+
+So far everything worked fine. What happens when there is a test failure? Let's
+create a faulty testcase to find out:
+
+.. code-block:: sh
+
+   $ mkdir multiplication
+   $ cd multiplication
+   $ touch test.yaml
+   $ echo "2 * 3" > input.bc
+   $ echo 8 > test.out
+   $ cd ..
+   $ ./testsuite.py
+   INFO     Found 3 tests
+   INFO     PASS         subtraction
+   INFO     PASS         addition
+   INFO     FAIL         multiplication: unexpected output
+   INFO     Summary:
+   _  PASS         2
+   _  FAIL         1
+
+Instead of the expected ``PASS`` test result, we have a ``FAIL`` one with a
+message: ``unexpected output``. Even though we can easily guess the error is
+that the expected output should be ``6`` (not ``8``), let's ask the testsuite
+to show details thanks to the ``--show-error-output/-E`` option. We'll also ask
+the testsuite to run only that failing testcase:
+
+.. code-block:: sh
+
+   $ ./testsuite.py -E multiplication
+   INFO     Found 1 tests
+   INFO     FAIL         multiplication: unexpected output
+   _--- expected
+   _+++ output
+   _@@ -1 +1 @@
+   _-8
+   _+6
+   INFO     Summary:
+   _  FAIL         1
+
+On baseline comparison failure, ``DiffTestDriver`` creates a unified diff
+between the baseline (``--- expected``) and the actual output (``+++ output``)
+showing the difference, and the testsuite's ``--show-error-output/-E`` option
+displays it, making it easy to quickly spot the difference between the two.
+
+Even though these 3 testcases take very little time to run, most testsuites
+require a lot of CPU time to run to completion. Nowadays, most working stations
+have several cores, so we can spawn one test per core to speedup testsuite
+execution time. ``e3.testsuite`` supports the ``--jobs/-j`` option to achive
+this. This option works the same way it does for the ``make`` program: ``-jN``
+is the default (run at most N testcases at a time, default is 1), and ``-j0``
+tells to set N to the number of CPU cores.
+
+
+Test execution control
+----------------------
+
+There is no obvious bug in ``bc`` that this documentation could expect to
+survive for long, so let's stick with this wrong ``multiplication`` testcase
+and pretend that ``bc`` should return ``8``. This is a known bug, and so the
+failure is expected for the time being. This situation occurs a lot in
+software: bugs often take a lot of time to fix, sometimes test failures come
+from bugs in upstream projects, etc.
+
+To keep testsuite reports readable/usable, it is convenient to tag failures
+that are temporarily accepted as ``XFAIL`` rather than ``FAIL``: the former
+is a failure that has been analyzed as acceptable for now, leaving the latter
+for unexpected regressions to investigate. Testcases using a driver that
+inherits from ``ClassicTestDriver`` can do that adding a ``control`` entry in
+their ``test.yaml`` file. To do that, append the following to
+``multiplication/test.yaml``:
+
+.. code-block:: yaml
+
+   control:
```

Comment:
Highlighting is bad, you should probably specify a language.

Reply:
Actually there is a language (yaml). Looks like an issue with GitHub’s highlighter. This should be fine on readthedocs (my final target).